server/matchmaker.go

```diff
@@ -51,36 +98,159 @@ func NewMatchmakerService(name string) *MatchmakerService {
+	if requiredCount == 0 {
+		return map[MatchmakerKey]*MatchmakerProfile{}
+	}
+
+	if requiredCount > len(candidates) {
+		return nil
+	}
+
+	keys := make([]MatchmakerKey, 0)
+	values := make([]*MatchmakerProfile, 0)
+	for key, value := range candidates {
+		keys = append(keys, key)
+		values = append(values, value)
+	}
+
+	for i := 0; i < len(keys); i++ {
+		s := values[i]
+		tempCandidates := make(map[MatchmakerKey]*MatchmakerProfile, 0)
+		for j := i + 1; j < len(keys); j++ {
+			p := values[j]
+			if m.checkFilter(s, p) && m.checkFilter(p, s) {
+				tempCandidates[keys[j]] = p
 			}
 		}
+
+		findCandidateResult := m.crossmatchCandidates(tempCandidates, requiredCount-1)
+		if findCandidateResult != nil {
+			findCandidateResult[keys[i]] = s
+			return findCandidateResult
+		}
 	}
-	if int64(len(selected)) == requiredCount-1 {
-		for mk, _ := range selected {
-			delete(m.values, mk)
+	return nil
+}
+
+func (m *MatchmakerService) checkFilter(requestProfile, queuedProfile *MatchmakerProfile) bool {
+	if queuedProfile.RequiredCount != requestProfile.RequiredCount {
+		return false
+	}
+
+	for filterName, filter := range requestProfile.Filters {
+		propertyValue := queuedProfile.Properties[filterName]
+		if propertyValue == nil {
+			return false
+		}
+
+		if filter.Type() == TERM {
+			termFilter := filter.(*MatchmakerTermFilter)
+			propertyTermList, ok := propertyValue.([]string)
+			if !ok {
+				return false
+			}
+
+			matchingTerms := m.intersection(termFilter.Terms, propertyTermList)
+			if len(matchingTerms) == 0 {
+				return false
+			}
+
+			if termFilter.AllTerms && len(matchingTerms) != len(termFilter.Terms) {
+				return false
+			}
+		} else if filter.Type() == RANGE {
+			rangeFilter := filter.(*MatchmakerRangeFilter)
+			propertyInt, ok := propertyValue.(int64)
+
+			if !ok || propertyInt < rangeFilter.LowerBound || propertyInt > rangeFilter.UpperBound {
+				return false
+			}
+		} else if filter.Type() == BOOL {
+			boolFilter := filter.(*MatchmakerBoolFilter)
+			propertyBool, ok := propertyValue.(bool)
+			if !ok || boolFilter.Value != propertyBool {
+				return false
+			}
 		}
-		selected[qmk] = qmp
-	} else {
-		m.values[qmk] = qmp
 	}
-	m.Unlock()
 
-	if int64(len(selected)) != requiredCount {
-		return ticket, nil
-	} else {
-		return ticket, selected
+	return true
+}
+
+func (m *MatchmakerService) calculateAcceptedProperties(matched map[MatchmakerKey]*MatchmakerProfile) []*MatchmakerAcceptedProperty {
+	props := make([]*MatchmakerAcceptedProperty, 0)
+	for key, profile := range matched {
+		prop := &MatchmakerAcceptedProperty{
+			UserID:     key.UserID,
+			Properties: profile.Properties,
+			Filters:    profile.Filters,
+		}
+		props = append(props, prop)
+	}
+
+	return props
+}
+
+func (m *MatchmakerService) intersection(a, b []string) []string {
+	o := make([]string, 0)
+	for i := range a {
+		for j := range b {
+			if a[i] == b[j] {
+				o = append(o, a[i])
```

Comment:
This algo is fine for now, but we can at least break out of the inner loop once we've found our item.